File: app/src/main/java/com/ivanovsky/passnotes/presentation/core/widget/ErrorPanelView.java
package com.ivanovsky.passnotes.presentation.core.widget;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.ivanovsky.passnotes.R;

public class ErrorPanelView extends LinearLayout {

	private View retryButton;
	private TextView errorTextView;

	public enum State {
		MESSAGE,
		MESSAGE_WITH_RETRY
	}

	public ErrorPanelView(Context context) {
		super(context);
		init();
	}

	public ErrorPanelView(Context context, AttributeSet attrs) {
		super(context, attrs);
		init();
	}

	private void init() {
		setBackgroundResource(R.color.material_error_panel_background);
		setOrientation(HORIZONTAL);

		LayoutInflater.from(getContext()).inflate(R.layout.view_error_panel, this, true);

		errorTextView = findViewById(R.id.text);
		retryButton = findViewById(R.id.retryButton);
	}

	public void setText(CharSequence text) {
		errorTextView.setText(text);
	}

	public void setText(int resId) {
		errorTextView.setText(resId);
	}

	public void setState(State state) {
		if (state == State.MESSAGE) {
			retryButton.setVisibility(View.GONE);
		} else if (state == State.MESSAGE_WITH_RETRY) {
			retryButton.setVisibility(View.VISIBLE);
		}
	}
}

File: app/src/main/kotlin/com/ivanovsky/passnotes/presentation/core/widget/ErrorPanelView.kt
package com.ivanovsky.passnotes.presentation.core.widget

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.FrameLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.isVisible
import com.ivanovsky.passnotes.data.entity.OperationError
import com.ivanovsky.passnotes.databinding.ViewErrorPanelBinding
import com.ivanovsky.passnotes.domain.ResourceProvider
import com.ivanovsky.passnotes.extensions.formatReadableMessage
import com.ivanovsky.passnotes.presentation.core.ScreenState
import com.ivanovsky.passnotes.presentation.core.dialog.reportErrorDialog.ReportErrorDialog
import com.ivanovsky.passnotes.presentation.core.dialog.reportErrorDialog.ReportErrorDialogArgs
import com.ivanovsky.passnotes.presentation.core.widget.entity.OnButtonClickListener
import com.ivanovsky.passnotes.util.StringUtils.EMPTY

class ErrorPanelView : FrameLayout {

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
    constructor(context: Context) : super(context, null)

    var state: ScreenState? = null
        set(value) {
            field = value
            applyState(value)
        }

    var actionButtonClickListener: OnButtonClickListener? = null

    private val resourceProvider = ResourceProvider(context)
    private val binding: ViewErrorPanelBinding
    private var currentError: OperationError? = null

    init {
        binding = ViewErrorPanelBinding.inflate(
            LayoutInflater.from(context),
            this,
            true
        )

        binding.actionButton.setOnClickListener { actionButtonClickListener?.onButtonClicked() }
        binding.reportButton.setOnClickListener { onReportButtonClicked() }
        binding.closeIcon.setOnClickListener { onCloseIconClicked() }
    }

    private fun applyState(screenState: ScreenState?) {
        if (screenState == null) return

        currentError = screenState.error

        binding.text.text = screenState.error?.formatReadableMessage(resourceProvider) ?: EMPTY
        binding.actionButton.text = screenState.errorButtonText
        binding.actionButton.isVisible = (!screenState.errorButtonText.isNullOrEmpty())

        binding.root.isVisible = screenState.isDisplayingErrorPanel
    }

    private fun onCloseIconClicked() {
        binding.root.isVisible = false
    }

    private fun onReportButtonClicked() {
        val error = currentError ?: return

        val fragmentManager = (context as AppCompatActivity).supportFragmentManager

        val dialog = ReportErrorDialog.newInstance(
            args = ReportErrorDialogArgs(error)
        )
        dialog.show(fragmentManager, ReportErrorDialog.TAG)
    }
}